### PR TITLE
multiple slackbot support

### PR DESCRIPTION
**Pull request by hagen-danswer**

### CODE REVIEW

File: backend/alembic/versions/4ee1287bd26a_add_multiple_slack_bot_support.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        "slack_channel_config",
+        sa.Column("id", sa.Integer(), nullable=False),
+        sa.Column("slack_bot_id", sa.Integer(), nullable=True),
+        sa.Column("persona_id", sa.Integer(), nullable=True),
+        sa.Column("channel_config", postgresql.JSONB(), nullable=False),
+        sa.Column("response_type", sa.String(), nullable=False),
+        sa.Column(
+            "enable_auto_filters", sa.Boolean(), nullable=False, server_default="false"
+        ),
+        sa.ForeignKeyConstraint(
```
Comment (rkuo-danswer) [inline backend/alembic/versions/4ee1287bd26a_add_multiple_slack_bot_support.py:54]: i think sqlalchemy and other generators have a common naming convention for foreign key constraints, we should follow that.

File: web/src/components/EditableTextAreaDisplay.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  const resetEditing = () => {
+    setIsEditing(false);
+    setEditableValue(value);
+  };
+
+  const handleKeyDown = (
+    e: React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>
+  ) => {
+    if (e.key === "Enter") {
```
Comment (rkuo-danswer) [inline web/src/components/EditableTextAreaDisplay.tsx:68]: Does this if statement make sense?

File: web/src/components/EditableStringFieldDisplay.tsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+  const resetEditing = () => {
+    setIsEditing(false);
+    setEditableValue(value);
+  };
+
+  const handleKeyDown = (
+    e: React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>
+  ) => {
+    if (e.key === "Enter") {
```
Comment (rkuo-danswer) [inline web/src/components/EditableStringFieldDisplay.tsx:69]: Does this if statement make sense?

File: web/src/components/BackButton.tsx [hunk 4]
```diff
@@ -27,6 +29,8 @@ export function BackButton({
       onClick={() => {
         if (behaviorOverride) {
           behaviorOverride();
+        } else if (routerOverride) {
```
Comment (rkuo-danswer) [inline web/src/components/BackButton.tsx:32]: is this the proper way to do this?

File: web/src/app/admin/documents/sets/page.tsx [hunk 5]
```diff
@@ -314,9 +314,9 @@ const Main = () => {
       <Text className="mb-3">
         <b>Document Sets</b> allow you to group logically connected documents
         into a single bundle. These can then be used as filter when performing
```
Comment (rkuo-danswer) [inline web/src/app/admin/documents/sets/page.tsx]: as "a" filter.

File: web/src/app/admin/documents/sets/page.tsx [hunk 6]
```diff
@@ -314,9 +314,9 @@ const Main = () => {
...
-        searches in the web UI or attached to slack bots to limit the amount of
-        information the bot searches over when answering in a specific channel
-        or with a certain command.
+        searches in the web UI or attached to Slack Channels to limit the amount
+        of information the bot searches over when answering in a specific
```
Comment (rkuo-danswer) [inline web/src/app/admin/documents/sets/page.tsx]: This text seems overly wordy but mostly not in scope.

File: web/src/app/admin/bots/new/page.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  return (
+    <div className="container mx-auto">
+      <BackButton />
+      <AdminPageTitle
+        icon={<SourceIcon iconSize={32} sourceType={"slack"} />}
+        title="Configure DanswerBot for Slack Channel"
+      />
+
+      <SlackChannelConfigCreationForm
+        slack_bot_id={slack_bot_id}
```
Comment (rkuo-danswer) [inline web/src/app/admin/bots/new/page.tsx]: naming style is inconsistent

File: backend/danswer/db/slack_channel_config.py [hunk 8]
```diff
@@ -22,8 +22,8 @@
 )
 
 
-def _build_persona_name(channel_names: list[str]) -> str:
-    return f"{SLACK_BOT_PERSONA_PREFIX}{'-'.join(channel_names)}"
+def _build_persona_name(channel_name: str) -> str:
+    return f"{SLACK_BOT_PERSONA_PREFIX}{channel_name}"
```
Comment (rkuo-danswer) [inline backend/danswer/db/slack_channel_config.py:26]: this was joining with a dash before, did we want to leave that out?

File: backend/danswer/server/manage/models.py [hunk 9]
```diff
@@ -183,25 +197,47 @@ class SlackBotConfig(BaseModel):
...
                 StandardAnswerCategory.from_model(standard_answer_category_model)
-                for standard_answer_category_model in slack_bot_config_model.standard_answer_categories
+                for standard_answer_category_model in slack_channel_config_model.standard_answer_categories
             ],
-            enable_auto_filters=slack_bot_config_model.enable_auto_filters,
+            enable_auto_filters=slack_channel_config_model.enable_auto_filters,
+        )
+
+
+class SlackBot(BaseModel):
```
Comment (rkuo-danswer) [inline backend/danswer/server/manage/models.py:223]: would be good to comment what this model is / used for

File: backend/danswer/danswerbot/slack/listener.py [hunk 10]
```diff
@@ -169,6 +175,56 @@ def heartbeat_loop(self) -> None:
...
 
+    def _manage_clients_per_tenant(
+        self, db_session: Session, tenant_id: str | None, bot: SlackBot
+    ) -> None:
+        slack_bot_tokens = SlackBotTokens(
+            bot_token=bot.bot_token,
+            app_token=bot.app_token,
+        )
+
+        logger.debug(
```
Comment (rkuo-danswer) [inline backend/danswer/danswerbot/slack/listener.py]: doesn't seem super useful even for debugging once the problem has been figured out.

File: backend/danswer/danswerbot/slack/listener.py [hunk 11]
```diff
@@ -169,6 +175,56 @@ def heartbeat_loop(self) -> None:
...
+            bot.id,
+        ) not in self.slack_bot_tokens or slack_bot_tokens != self.slack_bot_tokens[
+            (tenant_id, bot.id)
+        ]:
+            if (tenant_id, bot.id) in self.slack_bot_tokens:
+                logger.info(
+                    f"Slack Bot tokens have changed for tenant {tenant_id}, bot {bot.id} - reconnecting"
+                )
+            else:
+                search_settings = get_current_search_settings(db_session)
```
Comment (rkuo-danswer) [inline backend/danswer/danswerbot/slack/listener.py:207]: does the slack bot need to warm up the model server? I feel like the celery workers already do this.

File: backend/danswer/danswerbot/slack/listener.py [hunk 12]
```diff
@@ -169,6 +175,56 @@ def heartbeat_loop(self) -> None:
...
+            logger.debug(
+                f"No Slack bot token found for tenant {tenant_id}, bot {bot.id}"
+            )
+            if (tenant_id, bot.id) in self.socket_clients:
+                asyncio.run(self.socket_clients[tenant_id, bot.id].close())
+                del self.socket_clients[tenant_id, bot.id]
+                del self.slack_bot_tokens[tenant_id, bot.id]
+            return
+
+        if (
```
Comment (rkuo-danswer) [inline backend/danswer/danswerbot/slack/listener.py]: this if statement and nested ones are pretty complicated ... any way to simplify?

File: backend/danswer/danswerbot/slack/listener.py [hunk 13]
```diff
@@ -281,26 +301,37 @@ def send_heartbeats(self) -> None:
...
+        )
 
     def stop_socket_clients(self) -> None:
         logger.info(f"Stopping {len(self.socket_clients)} socket clients")
-        for tenant_id, client in self.socket_clients.items():
-            if client:
-                asyncio.run(client.close())
-                logger.info(f"Stopped SocketModeClient for tenant {tenant_id}")
+        for (tenant_id, slack_bot_id), client in self.socket_clients.items():
+            asyncio.run(client.close())
```
Comment (rkuo-danswer) [inline backend/danswer/danswerbot/slack/listener.py:325]: no if client needed here?

**Review by rkuo-danswer — CHANGES_REQUESTED** [review]

File: web/src/components/BackButton.tsx (hunk 4 above)
Comment (hagen-danswer) [reply web/src/components/BackButton.tsx:32]: chris seemed okay with it lol

File: web/src/app/admin/documents/sets/page.tsx (hunk 6 above)
Comment (hagen-danswer) [reply web/src/app/admin/documents/sets/page.tsx]: agreed! has been reduced

File: backend/danswer/db/slack_channel_config.py (hunk 8 above)
Comment (hagen-danswer) [reply backend/danswer/db/slack_channel_config.py:26]: yes, its just one name per now

File: backend/danswer/danswerbot/slack/listener.py (hunk 13 above)
Comment (hagen-danswer) [reply backend/danswer/danswerbot/slack/listener.py:325]: no

File: backend/danswer/danswerbot/slack/listener.py (hunk 12 above)
Comment (hagen-danswer) [reply backend/danswer/danswerbot/slack/listener.py]: I greatly simplified it, good thought

**Review by rkuo-danswer — APPROVED** [review]

**Pull request merged**